pkg/edit/store_api.go

```diff
@@ -10,37 +10,88 @@ import (
 	"src.elv.sh/pkg/eval"
 	"src.elv.sh/pkg/eval/vals"
 	"src.elv.sh/pkg/parse/parseutil"
+	"src.elv.sh/pkg/store"
 )
 
 var errStoreOffline = errors.New("store offline")
 
 //elvdoc:fn command-history
 //
-// Outputs the entire command history as a stream of maps. Each map has a `id`
-// key that identifies the sequence number of the entry, and a `cmd` key that
-// identifies the content.
+// ```elvish
+// edit:command-history &dedup &newest-first
+// ```
 //
-// Use indexing to extract individual entries. For example, to extract the
-// content of the last command, do this:
+// Outputs the command history as a stream of maps unless `&cmd-only` is true. Each map has a `id`
+// key for the sequence number of the command, and a `cmd` key for the text of the command. If
+// `&dedup` is true only the most recent instance of each command (when comparing just the `cmd`
+// key) is output. If `&newest-first` is true the output is in newest to oldest order (default is
+// oldest to newest). If `&cmd-only` is true then only the command text is output (default is a map
+// as described previously).
+//
+// Use indexing to extract individual entries. For example, to extract the text of the most recent
+// command do one of these:
 //
 // ```elvish
 // edit:command-history | put [(all)][-1][cmd]
+// edit:command-history &cmd-only &newest-first | take 1
 // ```
+//
+// @cf builtin:dir-history
 
-func commandHistory(fuser histutil.Store, ch chan<- interface{}) error {
+type cmdhistOpt struct{ CmdOnly, Dedup, NewestFirst bool }
+
+func (o *cmdhistOpt) SetDefaultOptions() {}
+
+func commandHistory(opts cmdhistOpt, fuser histutil.Store, ch chan<- interface{}) error {
 	if fuser == nil {
 		return errStoreOffline
 	}
 	cmds, err := fuser.AllCmds()
 	if err != nil {
 		return err
 	}
-	for _, cmd := range cmds {
-		ch <- vals.MakeMap("id", strconv.Itoa(cmd.Seq), "cmd", cmd.Text)
+	if opts.Dedup {
+		cmds = uniqCmds(cmds, opts.NewestFirst)
+	} else if opts.NewestFirst {
+		reverse(cmds)
+	}
+	if opts.CmdOnly {
+		for _, cmd := range cmds {
+			ch <- cmd.Text
+		}
+	} else {
+		for _, cmd := range cmds {
+			ch <- vals.MakeMap("id", strconv.Itoa(cmd.Seq), "cmd", cmd.Text)
+		}
 	}
 	return nil
 }
 
+func uniqCmds(allCmds []store.Cmd, newestFirst bool) []store.Cmd {
```

Comment:
1. Unnecessary for `uniqCmds` to take `newestFirst` as an argument; this merely duplicates the reverse in the previous function.

2. Since the option is called "dedup", better to be consistent and call this function `dedupCmds` too.

Reply:
Hmm, I realized that the reason you made this function take `newestFirst` as an argument is that it naturally reverses the commands, so I'll keep that.